restrict_download_region/restrict_region.py

```diff
@@ -0,0 +1,141 @@
+import os
+import logging
+import botocore
+import boto3
+import json
+import restrict_download_region.cfnresponse as cfnresponse
+from contextlib import contextmanager
+import urllib3
+from typing import List, Dict
+
+# AWS_REGION should always be defined in the context of a lambda
+AWS_REGION = os.environ.get('AWS_REGION')
+BUCKET_NAME = os.environ.get('BUCKET_NAME')
+
+POLICY_STATEMENT_ID = "DenyGetObjectForNonMatchingIp"
+SINGLE_REGION_BUCKET_TAG = 'single-region-access'
+
+
+def handler(event: dict, context: dict):
+    """
+    This lambda will either be triggered by a CloudFormation Custom Resource
+    creation or by a recurring SNS topic
+    """
+    # context manager for the case when this lambda is triggered by aws custom resource
+    with handle_custom_resource_status_message(event, context) as custom_resource_request_type:
+        if not BUCKET_NAME:
+            raise ValueError("BUCKET_NAME must be defined in enviroment variables")
+
+        region_ip_prefixes = get_ip_prefixes_for_region() \
+            if custom_resource_request_type != 'Delete'\
+            else None
+
+        s3_client = boto3.client('s3')
+
+        # get current bucket_policy from the s3 bucket
+        bucket_policy = get_bucket_policy(s3_client, BUCKET_NAME)
+
+        # add/update/remove ip restirction policy depending on whether region_ip_prefixes
+        process_ip_restrict_policy(BUCKET_NAME, region_ip_prefixes,
+                                   custom_resource_request_type, bucket_policy)
+
+        # update with newly modified bucket policy
+        update_bucket_policy(s3_client, BUCKET_NAME, bucket_policy)
+
+
+def generate_ip_address_policy(bucket_name: str, region_ip_prefixes: List[str]):
+    return {
+        'Sid': POLICY_STATEMENT_ID,
+        'Effect': 'Deny',
+        'Principal': '*',
+        'Action': 's3:GetObject',
+        'Resource': 'arn:aws:s3:::'+bucket_name+'/*',
+        'Condition': {
+            'NotIpAddress': {'aws:SourceIp': region_ip_prefixes},
```

Comment:
`region_ip_prefixes` can be None on a `Delete` event?  what gets returned in that case?

Reply:
`process_ip_restrict_policy` won't call this function if it's a `delete`